### ISSUE TITLE
Zod email validation blocks the main thread in version 3.22.2

### ISSUE THREAD

**Issue by SpongeBed81** [colinhacks/zod#2787]
Here is the version 3.21.4 of Zod:

https://github.com/colinhacks/zod/assets/56435044/0cb44a0c-90ac-486d-a435-a44590234c1b

And here is the version 3.22.2:

https://github.com/colinhacks/zod/assets/56435044/64ec1f54-3c08-4f7b-a159-96d4fc9a9510

here is the demo code:

```svelte
<script>
    import { z } from 'zod';

    let emailVal = '';

    let success;
    let counter = 0;

    $: {
        success = z.string().email().safeParse(emailVal).success;
        counter++;
    }
</script>

<input type="text" bind:value={emailVal} />
<p>success: {success} | {counter}</p>

```

and I also tried something like this:

```svelte
<script>
    import { z } from 'zod';
    import { onMount } from 'svelte';

    onMount(() => {
        console.log(z.string().email().safeParse('a'.repeat(50)).success);
    });
</script>
```

Both of them blocks the main thread and the second code doesn't even let me to see the page.

When I opened the performance tab in devtools I noticed it is doubling the time it takes the complete the task every time when I enter a new character to the input.

![chrome_yMIe5E2Et1](https://github.com/colinhacks/zod/assets/56435044/371295ce-0672-41b8-8f19-b41bf652bb0e)

| task | time     |
|------|----------|
| 1    | 10ms     |
| 2    | 19.33ms  |
| 3    | 38.42ms  |
| 4    | 75.17ms  |
| 5    | 151ms    |
| 6    | 299.96ms    |
| 7    | 598.56ms |
| 8    | 1.21s    |

**Comment by JulienZD** [comment]
It's vulnerable to ReDoS attacks, as also mentioned in [this issue](https://github.com/colinhacks/zod/issues/2609). The linked issue contains a workaround by manually defining a zod string with a regex to parse the email address.

**Comment by MrKoopie** [comment]
We can reproduce this error as well. Downgrading to 3.21.4 fixed it for us.

**Comment by MacsDickinson** [comment]
have a PR in for #2609 which should resolve this - https://github.com/colinhacks/zod/pull/2824

**Comment by SpongeBed81** [comment]
> have a PR in for #2609 which should resolve this - #2824

Thanks for your contribution. I will wait for your PR to get merged 😄

**Comment by SpongeBed81** [comment]
> We can reproduce this error as well. Downgrading to 3.21.4 fixed it for us.

It fixes the bug in the latest version but version 3.21.4 also has a bug that also blocks the main thread as I said. So, be careful 😉

**Comment by SpongeBed81** [comment]
> Does that PR address this [NVD](https://nvd.nist.gov/vuln/detail/CVE-2023-4316) ?

#2824 take a look at this. But the short answer is yes.

**Comment by FabioZangirolami** [comment]
Hello! I was facing a serious problem in my application with hook-form + zod + mui. After many test cases and frustrated days I found the problem and made 2 codesandbox:
the first one with zod 3.21.4 is working perfectly after leaving the email field with the following invalid email "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" the error warning is instantaneous! Now using zod 3.22.2 after the field loses focus, the process becomes very slow

https://codesandbox.io/s/hook-form-zod-mui-forked-5xmfhg zod 3.21.4
https://codesandbox.io/s/hook-form-zod-mui-forked-d2jhxk zod 3.22.2

**Comment by colinhacks** [comment]
Fixed by https://github.com/colinhacks/zod/pull/2824

Landed in Zod v3.22.3
